server/src/main/java/org/opensearch/common/blobstore/stream/read/listener/FileCompletionListener.java

```diff
@@ -0,0 +1,42 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.common.blobstore.stream.read.listener;
+
+import org.opensearch.core.action.ActionListener;
+
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Set;
+
+public class FileCompletionListener implements ActionListener<Integer> {
+    private final int numberOfParts;
+    private final String fileName;
+    private final Set<Integer> completedParts;
+    private final ActionListener<String> completionListener;
+
+    public FileCompletionListener(int numberOfParts, String fileName, ActionListener<String> completionListener) {
+        this.completedParts = Collections.synchronizedSet(new HashSet<>());
+        this.numberOfParts = numberOfParts;
+        this.fileName = fileName;
+        this.completionListener = completionListener;
+    }
+
+    @Override
+    public void onResponse(Integer partNumber) {
+        completedParts.add(partNumber);
+        if (completedParts.size() == numberOfParts) {
```

Comment:
I think you just need a thread safe counter here, right? Is it possible for the same part number to invoke `onResponse` more than once?

Also, this isn't really thread safe either due to the following scenario:

```
Thread 1: completedParts.add(1);
Thread 2: completedParts.add(2);
Thread 1: completedParts.size == numberOfParts evaluates to true, invokes completionListener.onResponse()
Thread 2: completedParts.size == numberOfParts evaluates to true, invokes completionListener.onResponse()
```

This should probably be either:


```
final boolean complete;
synchronized (completedParts) {
    completedParts.add(partNumber);
    complete = completedParts.size() == numberOfParts;
}
if (complete) {
    completionListener.onResponse(fileName);
}
```

or (if using an AtomicInteger):

```
if (completionCount.incrementAndGet() == numberOfParts) {
    completionListener.onResponse(fileName);
}
```